### ISSUE TITLE
RFC: Device Interface Functions

### ISSUE THREAD

**Issue by gkelly** [lowRISC/opentitan#1043]
The purpose of this RFC is to standardize the quality bar and general shape of Device Interface Functions: high-bar runtimless hardware abstractions used to actuate actual hardware.  Please see the [document here](https://docs.google.com/document/d/1rMDHLQXSDH8CaUg0cID5ARGxdbflV61uk4crKWjhxs0/edit#).

It's a distillation of work by @sjgitty @moidx @sriyerg @domrizz0 and @tjaychen.  In condensing it I tried to make sure I didn't over-constrain the requirements.  However, I might have missed some so please let me know in the doc.

If you'd like comment/suggest access to the doc please just ask and I'll grant it.  I didn't see a way to make the doc link-viewable but org-commentable.

**Comment by sjgitty** [comment]
reviewed. Thanks for authoring @gkelly, saved me one email. @sriyerg see if you can give it a look before you take a few weeks off

**Comment by silvestrst** [comment]
In section "Coverage Requirements":

**Comment by silvestrst** [comment]
Would it be useful to have some information in this specification about what fields must/should go into the state tracker structures?

For example:
bool timer_init(timer_t *timer, uintptr_t base_address);

"base_address" then can be stored in "timer" instance, I guess other useful fields could be "bool Initialised", etc...

Maybe even if we have a number of elements that would be applicable to all of the different peripheral DIFs, have a common structure (or something similar) that all of the individual DIF modules have to have in their state structure?

The above also made me think if it would be useful to have a common header file with common yet non hardware specific functionality that DIFs could make use of?

**Comment by silvestrst** [comment]
I think it could be useful to describe the verification extensions in the "Verification Stage Allowances" section, as the only place where these extensions are mentioned is at the bottom of the document when talking about directory layout, which I think is quite easy to miss (I have done first time I read the document).

**Comment by silvestrst** [comment]
In section "Coverage Requirements":
This is distinct from mandating that a DIF be required to cover all of the functionality of a given piece of hardware.

And

section "Source Layout":
They must comprise at least one header unique to the hardware in question

In the header file, do we want to have register definitions, etc. only required and used by the DIF, or we want full set of definitions? As mentioned in the "Coverage Requirements" section, the DIF does not necessarily exposes the full set of features provided by the hardware.

Is anything "extra" then up to the corresponding device driver to implement?

**Comment by sjgitty** [comment]
Silvestrs, those comments would be best in the document. I think you just
got comment access thanks
to Garret.

Scott

**Comment by silvestrst** [comment]
@sjgitty Thanks, I will move them to the document some time this weekend, or more likely on Monday

**Comment by mcy** [comment]
Some unresolved questions from reviewing other folks' DIF PRs, and from my own. I'm posting them here, but we should probably migrate them to the doc as feelings about them crystalize.
- In https://github.com/lowRISC/opentitan/issues/1224, @silvestrst brings up the testing issue, which @alphan and I have also been thinking about independently. There are two broad questions:
  - How are we doing unit tests? @alphan has been experimenting with gTest, but it's worth looking into other unit test frameworks for C/C++ (full disclosure, I'm a gTest fan).
  - Do we want to do "end to end" tests on real hardware? Compare the existing sanity tests under sw/device/tests.
- Should DIFs test preconditions, like arguments being within enum range and for null pointers? I'm fairly against this:
  - Callers should be responsible for not passing garbage to DIFs; a bad pointer to a machine register, or a pointer to unallocated memory, is insta-UB, but can't be checked as a precondition. C is an unsafe language, no matter what preconditions you check.
  - If a DIF returns an error due to such a programmer error, it's *still* the caller's problem to deal with it. Should they just explicitly crash? Seems like more work than the benefit that comes out.
- What should DIF functions return as an error? @silvestrst is using booleans as a stop-gap. I think we should consider returning enum values, like in Rust.
- We need a list of allowed headers in DIF libraries. Presumably, `stdasm.h` is now allowed, but `memcpy()` should be kosher. I'm also pushing to put other utilities, like `mmio.h` (aka `register.h`), and a "common enums" header, in libbase.

**Comment by lenary** [comment]
An opinion on checking preconditions:
- I do not think we should be reckless. C is unsafe, but we can still protect ourselves from its worst corners.
- Testing that pointer arguments are non-null is easy and doesn't take much code. The programmer is in a much better position if they get `false` back because the arguments were invalid, rather than maybe some of the memory writes working and some not, even if they do just have to hard crash.
- We should be checking that enums arguments are one of the enum variants, again, this is quite easy.
- I do agree that some harder preconditions (ensuring pointers are in specific memory ranges, or other complex preconditions) should not be checked.

**Comment by silvestrst** [comment]
Following up on yesterday's discussion, what there a consensus on enum range checking? For something like:
```
/**                                                                                                                                                                                                 
 * TX and RX FIFO depth configuration                                                                                                                                                               
 *                                                                                                                                                                                                  
 * Enumeration used to set the upper limit of bytes to queue.                                                                                                                                       
 */
typedef enum dif_uart_watermark {
  kDifUartWatermarkByte1 = 0, /**< 1 byte */
  kDifUartWatermarkByte4,     /**< 4 bytes */
  kDifUartWatermarkByte8,     /**< 8 bytes */
  kDifUartWatermarkByte16,    /**< 16 bytes */
  kDifUartWatermarkByte30,    /**< 30 bytes */
} dif_uart_watermark_t;
```

Do, or don't - something like this?:
```
// Validate the UART ON/OFF value                                                                                                                                                                   
static bool uart_watermark_valid(dif_uart_watermark_t val) {
  return (val >= kDifUartWatermarkByte1 || val <= kDifUartWatermarkByte30);
}
```

I don't have a strong opinion, maybe was being overly paranoid with range checking.

**Comment by lenary** [comment]
Yes, the consensus was don't do checking that allows every enum variant, only do the enum checking if `val` can only be a subset of the enum variants.

**Comment by silvestrst** [comment]
Cool, I will remove these checks from my PRs

**Comment by silvestrst** [comment]
I have a question regarding static functions documentation. I think it is useful to describe what static functions are doing, but I felt like full function boilerplate like we do for public functions might be overkill. I feel that something like comment style description would be sufficient for static functions. My approach is:
```
// Get IE, IP, LE register bit index from IRQ source ID. IRQ IDs start from 1,
// however IRQ bit index inside the registers start from 0. This means that in
// order to get IRQ bit index in the register, 1 has to be substracted from the
// obtained remainder.
static uint8_t plic_bit_index_from_irq_id(dif_plic_irq_src_id_t irq) {
  return ((uint32_t)irq % PLIC_ID_TO_INDEX_MODULO) - 1;
}
```
@alphan approach is to use public function format:
```
/**                                                                                                                                                                                                 
 * Gives the mask that corresponds to the given bit index. Calls |abort()|                                                                                                                          
 * if |index| is out of bounds for a 32-bit register.                                                                                                                                               
 *                                                                                                                                                                                                  
 * @param index Bit index in a 32-bit register.                                                                                                                                                     
 */
static inline uint32_t index_to_mask(const uint32_t index) {
  if (index >= (8 * sizeof(uint32_t))) {
    abort();
  }
  return 1U << index;
}
```
I don't know which comment style is more appropriate:
```
/**
* some text
* some text
*/
```
or
```
// some text
// some text
```
But I think that perhaps for static functions we shouldn't describe parameters (`@param index Bit index in a 32-bit register.`)?

What do you think?

**Comment by imphil** [comment]
You can use Doxygen-style comments without the `@param/@return` parts, so you don't need to choose a different comment style.

**Comment by lenary** [comment]
I don't see the value in having two different documentation comment formats. Doxygen seems able to omit generating docs for static functions in C (indeed seems to do this by defaut), so we shouldn't feel documenting static functions with a doxygen-style comment makes them "public".

I think you can choose to omit the `@param`/`@return` parts if you want, but in more complex static functions you should probably be using them.

**Comment by silvestrst** [comment]
Really what I wanted to figure out - how much private functions should be documented. @lenary I agree with what you said, and also I will change my PRs to also use Doxygen style description for private functions.

**Comment by imphil** [comment]
My general rule on that: Only document the ones which aren't obvious from the function name, e.g. have non-obvious side effects, pass on data ownership, do allocations, etc. Beyond that, use your intuition.

**Comment by silvestrst** [comment]
@imphil sounds good to me!

**Comment by silvestrst** [comment]
Hey everyone, I have another question regarding the static function parameters. Do we want to constify everything, like in @alphan approach:
```
void dif_gpio_set_output_modes_masked(const dif_gpio_t *const gpio,
                                      const uint32_t mask, const uint32_t val) {
  write_register_masked_access(gpio, GPIO_MASKED_OE_LOWER_REG_OFFSET,
                               GPIO_MASKED_OE_UPPER_REG_OFFSET, mask, val);
}
```

Or we only want to constify the data pointers point to (obviously only when required), like:
```
bool dif_uart_send_bytes(const dif_uart_t *uart, const uint8_t *data,
                         size_t bytes_requested, size_t *bytes_written) {
  if (uart == NULL || bytes_written == NULL) {
    return false;
  }

  *bytes_written = uart_send_bytes(uart, data, bytes_requested);

  return true;
}
```

Please let me know what you think.

**Comment by imphil** [comment]
Use const for pointers of course. 

Beyond that, I'm not a fan of things like `const uint32_t val`, I'd prefer `uint32_t val`. I don't see a benefit of adding `const` in such cases, beyond having to type more.

**Comment by lenary** [comment]
I agree with @imphil for pointers and sacalars. For struct parameters, `const` could be helpful, but they should all be passed with a pointer anyway.

**Comment by gkelly** [comment]
This guidance is aligned with the style guide, which strongly suggests const where it makes sense.  Pointers, yes.  Scalars, no.

**Comment by alphan** [comment]
I know I'm a little bit late but IMHO qualifying arguments that are passed by value as `const` has benefits. In the majority of the cases arguments to a function are not modified in the function body and `const T val` formally states that (and the compiler enforces it). You don't need to think "I see `val` here, but I wonder if it has been modified since we entered this function." Of course, this is more applicable to longer functions but I agree that it has no meaning in the function declaration. Is this something that we really need to enforce? Would the following be acceptable?

```
// Declaration in the header
bool dif_device_func(const dif_device_t * device, uint32_t addr);

// Definition
bool dif_device_func(const dif_device_t * const device, const uint32_t addr) {
   ...
}
```

**Comment by gkelly** [comment]
> We strongly recommend using const in APIs (i.e. on function parameters, methods, and non-local variables) wherever it is meaningful and accurate.

and

> Using const on local variables is neither encouraged nor discouraged.

cover this from the style guide I believe.  So I would say you're allowed to write the physically-const definition even though the declaration isn't.

**Comment by alphan** [comment]
Great! Thanks for the clarification!

**Comment by silvestrst** [comment]
I hate to come back to the const discussion, but can we just aim to have all the difs consistent in this regard.

> Using const on local variables is neither encouraged nor discouraged.

> cover this from the style guide I believe. So I would say you're allowed to write the physically-const definition even though the declaration isn't.

I guess the Google C++ Coding Style is not project specific, and hence gives some liberty. I think for DIFs, we should pick one method and stick to it for consistency. I don't have strong opinion one way or another, but I would like to be consistent.

**Comment by lenary** [comment]
My feeling on `const` is that scalar arguments should not have it, and that function declarations should match their definitions exactly. With this in mind, I have a note about `const` to match this in #1354. Let's move discussion of `const` there as it is not the most important part of the DIFs :)

**Comment by lenary** [comment]
The document was merged in #1915. I think this can be closed now?

**Comment by sjgitty** [comment]
I believe so too, please confirm and close @gkelly
